### PR TITLE
feat: Install keycloak for openshift using OpenShift templates

### PR DESCRIPTION
**Pull request by iam-veeramalla**
Signed-off-by: iam-veeramalla <abhishek.veeramalla@gmail.com>

**What type of PR is this?**
> /kind enhancement

**What does this PR do / why we need it**:
This PR adds `sso provider` in CR.   The reconciler installs Keycloack/RHSSO using OpenShift Templates.   This PR does not configure.  This PR also does not delete installed Keycloak.

**Have you updated the necessary documentation?**
Added

**Which issue(s) this PR fixes**:

Fixes #?
This PR address installing keycloak for openshift of #312

**How to test changes / Special notes to the reviewer**:
Deploy the example argocd CR
```
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
  labels:
    example: route
spec:
  sso:
    provider: keycloak
  server:
    route:
      enabled: true
```

or add the below section to your existing argocd CR Spec
```
  sso:
    provider: keycloak
```

* Get the SSO hostname or url. It can only be accessed using https://<hostname>

```
oc get route | grep sso
NAME                    HOST/PORT                                                                                  PATH   SERVICES                PORT    TERMINATION            WILDCARD
sso                     sso-default.apps.ci-ln-0ckf2vb-d5d6b.origin-ci-int-aws.dev.rhcloud.com                            sso                     <all>   reencrypt              None
```

* Get User and Password from the SSO pod name
```
oc exec sso-1-<> -- "env" | grep SSO_ADMIN_USERNAME
oc exec sso-1-<> -- "env" | grep SSO_ADMIN_PASSWORD
```

### CODE REVIEW

**Comment by wtam2018** [comment]
Please update user documents
v0.0.15 metadata should not be changed right?  It is being released.

File: pkg/apis/argoproj/v1alpha1/argocd_types.go [hunk 1]
```diff
@@ -419,6 +424,9 @@ type ArgoCDSpec struct {
 	// Server defines the options for the ArgoCD Server component.
 	Server ArgoCDServerSpec `json:"server,omitempty"`
 
+	// SSO defines the user management for Argo CD
```
Comment (wtam2018) [inline pkg/apis/argoproj/v1alpha1/argocd_types.go:427]: (Comments will go to the generated CRD as field documentation.)  SSO and user management are quite different.   Let's put more descriptive comment.

It should not be a pointer.

**Comment by iam-veeramalla** [comment]
> Please update user documents
> v0.0.15 metadata should not be changed right? It is being released.

@wtam2018 sorry, I did not get this

File: pkg/apis/argoproj/v1alpha1/argocd_types.go [hunk 2]
```diff
@@ -333,6 +333,11 @@ type ArgoCDServerServiceSpec struct {
 	Type corev1.ServiceType `json:"type"`
 }
 
+// ArgoCDSSOSpec defines the provider for user management on Argo CD.
```
Comment (wtam2018) [inline pkg/apis/argoproj/v1alpha1/argocd_types.go:336]: SSO != User management.

File: pkg/apis/argoproj/v1alpha1/argocd_types.go [hunk 3]
```diff
@@ -333,6 +333,11 @@ type ArgoCDServerServiceSpec struct {
...
+type ArgoCDSSOSpec struct {
+	Provider string `json:"provider,omitempty"`
```
Comment (wtam2018) [inline pkg/apis/argoproj/v1alpha1/argocd_types.go:338]: Please add comment to this feild.

File: pkg/controller/argocd/keycloak.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+	tmpl = template.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"description":               "RH-SSO Template for Gitops Operator",
+				"iconClass":                 "icon-sso",
+				"openshift.io/display-name": "Keycloak",
+				"tags":                      "keycloak",
+				"version":                   "9.0.4-SNAPSHOT",
+			},
+			Name:      "rhsso",
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:348]: Does the name has to be "rhsso"?

Comment (iam-veeramalla) [reply pkg/controller/argocd/keycloak.go:348]: Yes , since we have decided to use templates.

File: pkg/controller/argocd/sso.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	found, err := argoutil.VerifyAPI(template.SchemeGroupVersion.Group, template.SchemeGroupVersion.Version)
+	if err != nil {
+		return err
+	}
+	templateAPIFound = found
+	return nil
+}
+
+func (r *ReconcileArgoCD) reconcileSSO(cr *argoprojv1a1.ArgoCD) error {
+	if cr.Spec.SSO.Provider == "keycloak" {
```
Comment (wtam2018) [inline pkg/controller/argocd/sso.go:45]: cr.Spec.SSO can be nil, right?

File: pkg/controller/argocd/keycloak.go [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+func getKeycloakConfigMapTemplate(ns string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"description": "ConfigMap providing service ca bundle",
+				"service.beta.openshift.io/inject-cabundle": "true",
+			},
+			Labels: map[string]string{
+				"application": "${APPLICATION_NAME}",
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:51]: Where does ${APPLICATION_NAME} value gets populated?

File: pkg/controller/argocd/keycloak.go [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+			{Name: "DB_MIN_POOL_SIZE", Value: "${DB_MIN_POOL_SIZE}"},
+			{Name: "DB_MAX_POOL_SIZE", Value: "${DB_MAX_POOL_SIZE}"},
+			{Name: "DB_TX_ISOLATION", Value: "${DB_TX_ISOLATION}"},
+			{Name: "JGROUPS_PING_PROTOCOL", Value: "openshift.DNS_PING"},
+			{Name: "OPENSHIFT_DNS_PING_SERVICE_NAME", Value: "${APPLICATION_NAME}-ping"},
+			{Name: "OPENSHIFT_DNS_PING_SERVICE_PORT", Value: "8888"},
+			{Name: "X509_CA_BUNDLE", Value: "/var/run/configmaps/service-ca/service-ca.crt /var/run/secrets/kubernetes.io/serviceaccount/ca.crt"},
+			{Name: "JGROUPS_CLUSTER_PASSWORD", Value: "${JGROUPS_CLUSTER_PASSWORD}"},
+			{Name: "SSO_ADMIN_USERNAME", Value: "${SSO_ADMIN_USERNAME}"},
+			{Name: "SSO_ADMIN_PASSWORD", Value: "${SSO_ADMIN_PASSWORD}"},
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:73]: Cleartext passwords are viewable in the container's environment variables?

File: pkg/controller/argocd/keycloak.go [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+			{Name: "SSO_REALM", Value: "${SSO_REALM}"},
+			{Name: "SSO_SERVICE_USERNAME", Value: "${SSO_SERVICE_USERNAME}"},
+			{Name: "SSO_SERVICE_PASSWORD", Value: "${SSO_SERVICE_PASSWORD}"},
+		},
+		Image:           "${APPLICATION_NAME}",
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:78]: application name == image name?

File: pkg/controller/argocd/keycloak.go [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+				MountPath: "/var/run/configmaps/service-ca",
+				Name:      "service-ca",
+				ReadOnly:  true,
+			},
+		},
+	}
+}
+
+func getKeycloakDeploymentConfigTemplate(ns string) *appsv1.DeploymentConfig {
+	keycloakContainer := getKeycloakContainer()
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:141]: Do we need to pass ns to the container?

File: pkg/controller/argocd/keycloak.go [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+	route, err := json.Marshal(routeTemplate)
+	if err != nil {
+		return tmpl, err
+	}
+
+	tmpl = template.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"description":               "RH-SSO Template for Gitops Operator",
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:342]: Gitops Operator?

File: pkg/controller/argocd/keycloak.go [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+			{
+				Raw: json.RawMessage(service),
+			},
+			{
+				Raw: json.RawMessage(pingService),
+			},
+			{
+				Raw: json.RawMessage(route),
+			},
+		},
```
Comment (wtam2018) [inline pkg/controller/argocd/keycloak.go:367]: After reading these code, I am not sure if we really need template. :-)   What we are doing here is to create a configmap, deploymentConfig/deployment (you don't need deploymentConfig unless we want re-deploy if image changes), keycloak service, ping service, and a route.

**Comment by wtam2018** [comment]
> > Please update user documents
> > v0.0.15 metadata should not be changed right? It is being released.
> 
> @wtam2018 sorry, I did not get this

I see changes in deploy/olm-catalog/argocd-operator/0.0.15/argocd-operator.v0.0.15.clusterserviceversion.yaml  but @jomkz is releasing 0.0.15?   Do you mean to create v0.0.16 metadata?  Or, this PR is targeted for v0.0.15?

File: pkg/controller/argocd/keycloak.go (hunk 11 above)
Comment (iam-veeramalla) [reply pkg/controller/argocd/keycloak.go:367]: I  raised this question in the last design meeting. I understood that we need this as this is supported way of deployment and support team is comfortable with this.

Also, deployment config comes with image streams which is an added advantage.

File: pkg/controller/argocd/sso.go (hunk 5 above)
Comment (iam-veeramalla) [reply pkg/controller/argocd/sso.go:45]: reconcileSSO is only reconciled when cr.Spec.SSO != nil

util.go
-----
	if cr.Spec.SSO != nil {
		log.Info("reconciling SSO")
		if err := r.reconcileSSO(cr); err != nil {
			return err
		}
	}

File: pkg/controller/argocd/keycloak.go (hunk 7 above)
Comment (iam-veeramalla) [reply pkg/controller/argocd/keycloak.go:73]: yes the SSO_ADMIN_PASSWORD is viewable in the container's env variable.

oc exec sso-1-<> -- "env" | grep SSO_ADMIN_PASSWORD
---------------------------------------
SSO_ADMIN_PASSWORD=cTWjSwxS

File: pkg/controller/argocd/keycloak.go (hunk 8 above)
Comment (iam-veeramalla) [reply pkg/controller/argocd/keycloak.go:78]: Image stream sets the Image to "sso74-openshift-rhel8:7.4"

**Comment by iam-veeramalla** [comment]
> > > Please update user documents
> > > v0.0.15 metadata should not be changed right? It is being released.
> > 
> > 
> > @wtam2018 sorry, I did not get this
> 
> I see changes in deploy/olm-catalog/argocd-operator/0.0.15/argocd-operator.v0.0.15.clusterserviceversion.yaml but @jomkz is releasing 0.0.15? Do you mean to create v0.0.16 metadata? Or, this PR is targeted for v0.0.15?

gotcha !!

File: docs/reference/argocd.md [hunk 12]
```diff
@@ -1049,7 +1074,7 @@ spec:
   usersAnonymousEnabled: false
 ```
 
-## Version
+x## Version
```
Comment (wtam2018) [inline docs/reference/argocd.md:1077]: remove "x" character.

File: docs/reference/argocd.md [hunk 13]
```diff
@@ -999,6 +1000,30 @@ spec:
   statusBadgeEnabled: true
 ```
 
+## Single sign-on Options
+
+The following properties are available for configuring the Single sign-on component.
+
```
Comment (wtam2018) [inline docs/reference/argocd.md:1006]: please add a note that currently only openshift is supported for keycloak provider.

File: docs/usage/keycloak.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
+# Usage
+
+The following example shows the most minimal valid manifest to create a new Argo CD cluster with keycloak as Single sign-on provider.
+
```
Comment (wtam2018) [inline docs/usage/keycloak.md:4]: please add a note that currently only openshift is supported for keycloak provider.

File: docs/usage/keycloak.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+Create a new Argo CD cluster in the `argocd` namespace using the provided basic example.
+
+```bash
+kubectl create -n argocd -f examples/argocd-basic.yaml
+```
+
+The above configuration creates a keycloak instance and its relevant resources along with the Argo CD resources. Users can login into the keycloak console using the below commands.
+
+Get the Keycloak Route URL for Login.
+
```
Comment (wtam2018) [inline docs/usage/keycloak.md:28]: Route is openshift specific.   I wonder if we create an ingress for kubernetes clusters?

File: docs/usage/keycloak.md (hunk 14 above)
Comment (iam-veeramalla) [reply docs/usage/keycloak.md:6]: added

File: docs/usage/keycloak.md (hunk 15 above)
Comment (iam-veeramalla) [reply docs/usage/keycloak.md:30]: updated the note and updated the doc for openshift only. we will create a different section for k8s as we use ingress in that case and also usage steps might be slightly different.

File: docs/reference/argocd.md (hunk 12 above)
Comment (iam-veeramalla) [reply docs/reference/argocd.md]: removed :)

**Review by wtam2018 — APPROVED** [review]
LGTM
@iam-veeramalla thanks!

**Pull request merged**